### PR TITLE
Fix Issue 239 (modality sorting)

### LINKED ISSUE #239: GLM for SPS in GUI batch
### Summary
(same information as I mentioned in Slack)

in the batch, it asked us to specify which eye we want to analyse: best eye, last left, or last right;
however, in the script, the function is not really taking any information about which eye to use
so the "job" contains channel name as "sps" only, but not "sps_l" or "sps_r"
while in the data file to be analysed, the sps data was stored in "sps_l" and "sps_r"
here is the error message:

> ------------------------------------------------------------------------
>15-Jun-2021 16:49:28 - Running job #1
>------------------------------------------------------------------------
>15-Jun-2021 16:49:28 - Running 'GLM for SPS'
>Computing GLM: D:\Gitlab\yanfang_phd\PhD_projects\PsPM_workshop\2021_ESCAN2021\example_data\mdl_sps_DoxMem2_09_sn3_blk1.mat ...
>Getting data ...251             warning('ID:non_existing_channeltype', 'There are no channels of type ''%s'' in the datafile', chan); >return;
>Warning: There are no channels of type 'sps' in the datafile 
 > In pspm_load_data (line 251)
 > In pspm_glm (line 295)
>  In pspm_cfg_run_glm_sps (line 40)
>  In cfg_run_cm (line 29)
>  In cfg_util>local_runcj (line 1708)
>  In cfg_util (line 968)
>  In cfg_ui>MenuFileRun_Callback (line 703)
>  In gui_mainfcn (line 95)
>  In cfg_ui (line 53) 
>15-Jun-2021 16:50:19 - Done    'GLM for SPS'
>15-Jun-2021 16:50:19 - Done
> 

in glm line 295, it called pspm_load_data, which has "chan" as input argument, "chan" = "sps"
but this is wrong... "chan" should be "sps_l" or "sps_r" according to "best eye" or "left" "right" selected in the batch
In pspm_cfg_run_glm_sps (line 40), it called glm with "job"... so probably "job" did not really consider the selection of which eye to use
I suppose this should be easy to fixate? just according to which eye used, add the suffix to the "chan"?

here is a screenshot for the batch setting
image

### Technical Info
* PsPM revision ID: 5.1.0 beta
* MATLAB version: 2018b

### PR DESCRIPTION
**Pull request by teddychao**
Fixes #239 .

Changes proposed in this pull request:
- Update the two scripts for handling only the modality _sps_.

*pspm_cfg_glm_sps* and *pspm_cfg_run_glm_sps* should deal with the modality *sps*, thus the returned value for channels would be *sps* instead of *pupil*. *Pupil* is specified for the modality *psr*.

Thanks for the very helpful [comments](https://github.com/bachlab/PsPM/pull/241#pullrequestreview-685128037) by @SerenXia .

### CODE REVIEW

File: src/pspm_cfg/pspm_cfg_glm_sps.m [hunk 1]
```diff
@@ -77,19 +77,19 @@
 chan_def_left         = cfg_const;
 chan_def_left.name    = 'Last left eye';
 chan_def_left.tag     = 'chan_def_left';
-chan_def_left.val     = {'pupil_l'};
+chan_def_left.val     = {'sps_l'};
 chan_def_left.help    = {'Use last left eye channel.'};
```
Comment (SerenXia) [inline src/pspm_cfg/pspm_cfg_glm_sps.m:86]: could you please change the help text as well? 
'use the last sps channel from left eye'

**Comment by teddychao** [comment]
@SerenXia  Thank you for your suggestions for updating the descriptions.

I have a question regarding the _model.channel_. As you suggest, for the situation where the modality is *sps*, the returned _model.channel_ should be among *sps_l*, *sps_r*, and *sps*. I agree with this.

However, this is a confusing sentence in *pspm_glm*, which is at line 227-228
>  if ~isfield(model, 'channel')
> model.channel = model.modality;

I guess for the modality *sps*, this is applicable. However for *psr*, the returned values should be, based on my understanding, *pupil*. This is a little conflicting. Could I know *pspm_glm* is generic and should handle both *psr* and *sps*? If so, this sentence should be updated too.

Thanks.

**Review by dominikbach — APPROVED** [review]

**Review by SerenXia — COMMENTED** [review]
Hi @teddychao 
I got a bug again when testing the current fixed version.

> Warning: There are no channels of type 'sps' in the datafile 
> In pspm_load_data (line 251)
>   In pspm_glm (line 295)
>   In pspm_cfg_run_glm_sps (line 48)
>   In cfg_run_cm (line 29)
>   In cfg_util>local_runcj (line 1708)
>   In cfg_util (line 968)
>   In cfg_ui>MenuFileRun_Callback (line 703)
>   In gui_mainfcn (line 95)
>   In cfg_ui (line 53) 

This time the issue is here in the script pspm_load_data
from line 237 - line 239
`        if strcmpi(chan, 'pupil') && isfield(infos.source, 'best_eye')
            flag = get_chans_to_load_for_pupil(data, flag, infos.source.best_eye);
        else`

so we need another case, if chan is 'sps' and 'best_eye' selected
flag = get_chans_to_load_for_sps(data, flag, infos.source.best_eye)

you can simply copy this function from the function get_chans_to_load_for_pupil and change most of 'pupil' to 'sps'.
Just that, unlike pupil, sps doesn't have preprocessed channels... but if you remain the preprocess options, it should not hurt...

this subfunction is imbed in the script pspm_load_data


and some minor changes as I commented below

File: src/pspm_cfg/pspm_cfg_glm_sps.m [hunk 2]
```diff
@@ -77,19 +77,19 @@
...
-chan_def_right.val     = {'pupil_r'};
+chan_def_right.val     = {'sps_r'};
 chan_def_right.help    = {'Use last right eye channel.'};
 
 best_eye                = cfg_const;
 best_eye.name           = 'Best eye';
 best_eye.tag            = 'best_eye';
-best_eye.val            = {'pupil'};
+best_eye.val            = {'sps'};
 best_eye.help           = {['Use eye with the fewest NaN values.']};
```
[the remark this answers is not in the document]
Comment (teddychao) [reply src/pspm_cfg/pspm_cfg_glm_sps.m:98]: Yes done now, thanks for the suggestions 🙂

**Comment by SerenXia** [comment]
> @SerenXia Thank you for your suggestions for updating the descriptions.
> 
> I have a question regarding the _model.channel_. As you suggest, for the situation where the modality is _sps_, the returned _model.channel_ should be among _sps_l_, _sps_r_, and _sps_. I agree with this.
> 
> However, this is a confusing sentence in _pspm_glm_, which is at line 227-228
> 
> > if ~isfield(model, 'channel')
> > model.channel = model.modality;
> 
> I guess for the modality _sps_, this is applicable. However for _psr_, the returned values should be, based on my understanding, _pupil_. This is a little conflicting. Could I know _pspm_glm_ is generic and should handle both _psr_ and _sps_? If so, this sentence should be updated too.
> 
> Thanks.

@teddychao 
yes, pspm_glm is generic and should handle not only sps, psr, but also other modalities, like, hpr, rar, scr, sebr... etc.

**Comment by teddychao** [comment]
> > @SerenXia Thank you for your suggestions for updating the descriptions.
> > I have a question regarding the _model.channel_. As you suggest, for the situation where the modality is _sps_, the returned _model.channel_ should be among _sps_l_, _sps_r_, and _sps_. I agree with this.
> > However, this is a confusing sentence in _pspm_glm_, which is at line 227-228
> > > if ~isfield(model, 'channel')
> > > model.channel = model.modality;
> > 
> > 
> > I guess for the modality _sps_, this is applicable. However for _psr_, the returned values should be, based on my understanding, _pupil_. This is a little conflicting. Could I know _pspm_glm_ is generic and should handle both _psr_ and _sps_? If so, this sentence should be updated too.
> > Thanks.
> 
> @teddychao
> yes, pspm_glm is generic and should handle not only sps, psr, but also other modalities, like, hpr, rar, scr, sebr... etc.

Ok. I need to put some patch here.. I am not sure the default value should be the same to the modality name all the time. I will put the default value as _pupil_ for _psr_ modality, but the same to the modality name in other cases. It is marked in case there are potential bugs for other modalities in the future..

**Review by SerenXia — APPROVED** [review]
the current version worked :)

**Comment by teddychao** [comment]
> the current version worked :)

Thank you Yanfang!

**Pull request merged**